Reimplement as_strided in ATen.

zou3519 opened the pull request:
This moves away from using tensor.set_(...) for as_strided, which went
through TH and was weirdly slow/complicated. The new as_strided has a
new invariant that it will never resize the storage to a larger size
(the previous as_strided allowed that behavior but it seemed weird and
none of our code relied on it.)

This offers a small speedup on as_strided: it went from 1300ns to
1100ns although the benchmarks get a little noisy here.

Also on the changelog is a quick fix to resize_ code to avoid unsigned
underflow. I'll rewrite the resize_ zero dim logic in a future diff, it
doesn't make sense the way it is written right now.

Test Plan: ran tests

zou3519 committed:
Reimplement as_strided in ATen.

This moves away from using tensor.set_(...) for as_strided, which went
through TH and was weirdly slow/complicated. The new as_strided has a
new invariant that it will never resize the storage to a larger size
(the previous as_strided allowed that behavior but it seemed weird and
none of our code relied on it.)

This offers a small speedup on as_strided: it went from 1300ns to
1100ns although the benchmarks get a little noisy here.

Also on the changelog is a quick note to resize_ code to fix the
behavior in the future: it doesn't make sense the way it is written right now.

Test Plan: ran tests

ezyang commented on aten/src/ATen/native/Resize.h:
```diff
@@ -30,11 +30,13 @@ inline TensorImpl* resize_impl_cpu_(
     // NB: storage size can be different from numel.
     for (size_t dim = 0; dim < size.size(); ++dim) {
+      // FIXME: Don't rely on storage_size being negative...
+      // This behavior was carried over from TH
```
I want to point out that in chat, you asked a different question:

> is signed integer overflow undefined behavior?

Signed integer overflow is indeed undefined behavior, but **unsigned** integer overflow is well defined, and `size_t` is an unsigned type. Maybe you know this already, in that case, ignore me :)

zou3519 replied:
Ah yes I asked the wrong question. I'll change the sizes back to size_t then.

I think the comments still apply  because on a tensor with 0 dimensions each dimension could have stride 0 and we could have positive storage_offset `N`. In this case, this code would resize the storage to size `N`, which is not what we want.

ezyang commented on aten/src/ATen/native/Resize.h:
```diff
@@ -46,4 +48,58 @@ inline TensorImpl* resize_impl_cpu_(
+ * (size, stride, storageOffset) must be in bounds for the new storage.
+ */
+inline void setStorage(
+    TensorImpl* self,
```
nit: Consider taking `const Tensor&` here, and getting the storage impl inside this function.

ezyang commented on aten/src/ATen/native/Resize.h:
```diff
@@ -46,4 +48,58 @@ inline TensorImpl* resize_impl_cpu_(
+ */
+inline void setStorage(
+    TensorImpl* self,
+    StorageImpl* new_storage,
```
Ditto with Storage

ezyang commented on aten/src/ATen/native/Resize.h:
```diff
@@ -46,4 +48,58 @@ inline TensorImpl* resize_impl_cpu_(
+  AT_CHECK(old_storage, "Tensor: invalid null storage");
+  if (old_storage != new_storage) {
+    c10::raw::intrusive_ptr::incref(new_storage);
+    THTensor_stealAndSetStoragePtr(self, new_storage);
```
I think this code can be written more safely. Here's some context that will help:

1. `THTensor_stealAndSetStoragePtr` purely exists as a BC mechanism for old-style TH calls. Do not feel obligated to use it. And indeed, in this case, using it is inappropriate.
2. `storage_` on `TensorImpl` is currently public. It shouldn't be. One big reason it's public is to support the implementation of `THTensor_stealAndSetStoragePtr` which needs some sort of way of referring to the mutable reference thingy.

FYI! My recommendation would be to do something like add a new method to `TensorImpl` to do this operation directly on the smart pointer, rather than dealing with the raw pointers. This will be much safer, and much cleaner!

zou3519 replied:
That sounds reasonable, thank you for the suggestion :)

ezyang reviewed (changes requested):
Shout if you need this to land to unblock; I think this is OK as is, but can be even better!!

zou3519 committed:
Address comments

ezyang commented on aten/src/ATen/core/TensorImpl.h:
```diff
@@ -724,6 +724,17 @@ struct CAFFE2_API TensorImpl : public c10::intrusive_ptr_target {
+    if (new_storage_ == old_storage_) {
+      return;
+    }
+    storage_ = new_storage;
```
Huh. I wonder if `intrusive_ptr` assignment does the same-identity optimization.

ezyang replied:
Nope we don't. @smessmer, this seems worthwhile, right? Saves you a refcount twiddle.

smessmer replied:
I'm not sure it's worth it. Self assignment likely happens rarely and adding a branch has a cost for all calls, not just for the self-assignment calls. The branch predictor (at least on x86) is probably good enough to optimize for the non-self-assignment case, but then once a self-assignment happens, the cost of the branch misprediction might well be larger than the cost of atomically incrementing and decrementing the refcounter.

ezyang reviewed (approved):
Much better, thank you!

zou3519 committed:
Remove as_strided device guard.

as_strided calls torch.empty(..., self.options()) which uses its own
guard. The rest of the as_strided logic does not do any resizing of
storages or cuda kernel launches so it does not need a DeviceGuard.
